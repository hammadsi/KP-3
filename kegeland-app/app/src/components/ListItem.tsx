import React from 'react';
import {StyleProp, TextStyle} from 'react-native';
import {List, useTheme} from 'react-native-paper';

import Icon from './Icon';
import ListItemSkeleton from './ListItemSkeleton';

export type ListItemProps = Omit<
  React.ComponentProps<typeof List.Item>,
  'onLongPress'
> & {
  icon?: string;
  iconSize?: number;
  iconStyle?: StyleProp<TextStyle>;
  isRoute?: boolean;
  loading?: boolean;
  onLongPress?: () => void;
  render?: (props: {
    color: string;
    style?:
      | {
          marginRight: number;
          marginVertical?: number | undefined;
        }
      | undefined;
  }) => React.ReactNode;
};

/**
 * Component for rendering a list item.
 * @param props the props
 * @see {@link ListItemProps}
 * @see {@link List.Item}
 */
const ListItem: React.FC<ListItemProps> = ({
  icon,
  iconSize,
  iconStyle,
  isRoute,
  loading,
  onLongPress,
  render,
  ...props
}) => {
  const {colors} = useTheme();
  const handleOnPress =
    props.onPress || onLongPress
      ? () => {
          if (onLongPress) return () => null;
          return props.onPress && props.onPress();
        }
      : undefined;

  const handleOnLongPress = onLongPress
    ? () => {
        return onLongPress && onLongPress();
      }
    : undefined;

  return loading ? (
    <ListItemSkeleton />
  ) : (
    <List.Item
      {...props}
      onPress={handleOnPress}
      onLongPress={handleOnLongPress}
      delayLongPress={500}
      left={
        icon
          ? () => (
              <Icon
                color={colors.primary}
                icon={icon as string}
                size={iconSize}
                style={iconStyle}
              />
            )
          : undefined
      }
      title={props.title}
      right={
        isRoute
          ? ({color}) => <Icon color={color} icon="chevron-right" />
          : render
      }
    />
  );
};

ListItem.defaultProps = {
  iconSize: 24,
};

export default ListItem;
