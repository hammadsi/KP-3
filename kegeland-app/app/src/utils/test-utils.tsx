/* istanbul ignore file */
import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {Provider as PaperProvider} from 'react-native-paper';
import {render, RenderOptions} from '@testing-library/react-native';
import {Provider as ReduxProvider} from 'react-redux';

import {store} from '~state/store';
import {lightTheme} from '~constants/theme';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';

type TestWrapperProps = {
  children: React.ReactNode;
  store?: typeof store;
};

/**
 * Wrapper component with redux state. Used when testing components
 * which requires redux.
 * @param param0 test wrapper props @see {@link TestWrapperProps}
 * @returns
 */
export const TestWrapper: React.FC<TestWrapperProps> = ({
  children,
  ...props
}) => {
  return (
    <ReduxProvider store={props.store || mockStore(initialStore)}>
      <PaperProvider theme={lightTheme}>
        <NavigationContainer theme={lightTheme}>{children}</NavigationContainer>
      </PaperProvider>
    </ReduxProvider>
  );
};

/**
 * Custom test renderer which wraps components with a redux store provider
 * @param ui the component to render
 * @param options renderer options
 * @returns test rendered componnet
 */
export const customRender = <T extends any>(
  ui: React.ReactElement<T>,
  options?: {
    wrapperProps?: {
      store: any;
    };
    renderOptions?: RenderOptions;
  },
) =>
  render(ui, {
    wrapper: (props) => <TestWrapper {...props} {...options?.wrapperProps} />,
    ...options?.renderOptions,
  });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
