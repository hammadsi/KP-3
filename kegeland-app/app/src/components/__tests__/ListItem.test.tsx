import React from 'react';

import {render} from '~utils/test-utils';

import ListItem, {ListItemProps} from '../ListItem';

describe('Test ListItem-component', () => {
  it('should render correctly', () => {
    const props: ListItemProps = {
      title: 'Test',
    };
    const component = <ListItem {...props} />;
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should render skeleton on loading', () => {
    const props: ListItemProps = {
      title: 'Test',
      loading: true,
    };
    const component = <ListItem {...props} />;
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
