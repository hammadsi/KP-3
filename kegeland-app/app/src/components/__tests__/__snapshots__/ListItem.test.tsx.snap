// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ListItem-component should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        false,
        Array [
          Object {
            "padding": 8,
          },
          undefined,
        ],
      ]
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "marginVertical": 6,
              "paddingLeft": 8,
            },
            Object {
              "flex": 1,
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "fontSize": 16,
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                },
                undefined,
              ],
            ]
          }
        >
          Test
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`Test ListItem-component should render skeleton on loading 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "marginVertical": 8,
        "padding": 8,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 11,
          "height": 22,
          "opacity": 0.1,
          "width": 22,
        }
      }
    >
      <View
        style={Object {}}
      />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#000000",
          "borderRadius": 0,
          "flex": 1,
          "height": 22,
          "marginLeft": 8,
          "opacity": 0.1,
          "width": undefined,
        }
      }
    >
      <View
        style={Object {}}
      />
    </View>
  </View>
</View>
`;
